pallets/external-validator-slashes/src/lib.rs

```diff
@@ -0,0 +1,501 @@
+                let index = (index as usize) - removed;
+                era_slashes.remove(index);
+            }
+
+            // insert back slashes
+            Slashes::<T>::insert(&era, &era_slashes);
+            Ok(())
+        }
+
+        #[pallet::call_index(1)]
+        #[pallet::weight(T::WeightInfo::force_inject_slash())]
+        pub fn force_inject_slash(
+            origin: OriginFor<T>,
+            era: EraIndex,
+            validator: T::AccountId,
+            percentage: Perbill,
+        ) -> DispatchResult {
+            ensure_root(origin)?;
+            let active_era = T::EraIndexProvider::active_era().index;
+
+            ensure!(era <= active_era, Error::<T>::ProvidedFutureEra);
+
+            let slash_defer_duration = T::SlashDeferDuration::get();
+
+            let _ = T::EraIndexProvider::era_to_session_start(era)
+                .ok_or(Error::<T>::ProvidedNonSlashableEra)?;
+
+            let next_slash_id = NextSlashId::<T>::get();
+
+            let slash = compute_slash::<T>(
+                percentage,
+                next_slash_id,
+                era,
+                validator,
+                slash_defer_duration,
+            )
+            .ok_or(Error::<T>::ErrorComputingSlash)?;
+
+            // If we defer duration is 0, we immediately apply and confirm
+            let era_to_consider = if slash_defer_duration == 0 {
+                era
+            } else {
+                era.saturating_add(slash_defer_duration)
+                    .saturating_add(One::one())
+            };
+
+            let mut era_slashes = Slashes::<T>::get(&era_to_consider);
+            era_slashes.push(slash);
+
+            Slashes::<T>::insert(era_to_consider, &era_slashes);
+
+            NextSlashId::<T>::put(next_slash_id.saturating_add(One::one()));
+            Ok(())
+        }
+    }
+}
+
+/// This is intended to be used with `FilterHistoricalOffences`.
+impl<T: Config>
+    OnOffenceHandler<T::AccountId, pallet_session::historical::IdentificationTuple<T>, Weight>
+    for Pallet<T>
+where
+    T: Config<ValidatorId = <T as frame_system::Config>::AccountId>,
+    T: pallet_session::Config<ValidatorId = <T as frame_system::Config>::AccountId>,
+    T: pallet_session::historical::Config,
+    T::SessionHandler: pallet_session::SessionHandler<<T as frame_system::Config>::AccountId>,
+    T::SessionManager: pallet_session::SessionManager<<T as frame_system::Config>::AccountId>,
+    <T as pallet::Config>::ValidatorIdOf: Convert<
+        <T as frame_system::Config>::AccountId,
+        Option<<T as frame_system::Config>::AccountId>,
+    >,
+{
+    fn on_offence(
+        offenders: &[OffenceDetails<
+            T::AccountId,
+            pallet_session::historical::IdentificationTuple<T>,
+        >],
+        slash_fraction: &[Perbill],
+        slash_session: SessionIndex,
+    ) -> Weight {
+        let mut consumed_weight = Weight::from_parts(0, 0);
```

Comment:
Can we avoid the need to return a weight from this function? Not a fan of manually counting storage reads and writes.

Reply:
I can check but if it is meaningful to be used (i.e., ther weight charged when you insert an offence) then there is not much we can do. At best I can try to benchmark it, would that be better?

Reply:
Nice so let's remove it. And no need to benchmark this here, the weight will be included in the benchmark of whatever calls `on_offence` (right?).

Reply:
yes in theory yes (if those extrinsics have benchmarks)

Reply:
https://github.com/paritytech/polkadot-sdk/blob/b326540184a1c87c141c169896c2f0e58d41e24f/polkadot/runtime/rococo/src/lib.rs#L368

Those pallets use the default weightInfo in polkadot runtimes (and so are we)

Reply:
so I will remove it for now, but we need to track whether we need to track whether we need to properly benchmark those extrinsics

Reply:
fixed